onnxruntime/core/providers/webnn/builders/impl/shape_op_builder.cc

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Intel Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/common.h"
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/webnn/builders/helper.h"
+#include "core/providers/webnn/builders/model_builder.h"
+#include "core/providers/webnn/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace webnn {
+
+class ShapeOpBuilder : public BaseOpBuilder {
+  // Add operator related.
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related.
+ private:
+  bool IsOpSupportedImpl(const InitializedTensorSet& /* initializers */, const Node& node,
+                         const WebnnDeviceType device_type, const logging::Logger& logger) const override;
+};
+
+Status ShapeOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                             const Node& node,
+                                             const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  std::vector<int64_t> input_shape;
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[0], input_shape, logger), "Cannot get shape");
+  const auto rank = static_cast<int32_t>(input_shape.size());
+
+  NodeAttrHelper helper(node);
+  auto true_start = helper.Get("start", 0);
+  auto true_end = helper.Get("end", rank);
+
+  // Deal with negative(s) and clamp.
+  true_start = true_start < 0 ? true_start + rank : true_start;
+  true_start = true_start < 0 ? 0 : ((true_start > rank) ? rank : true_start);
```

Comment:
I recall you had a helper to handle negative axes, but I suppose since you're also clamping, that's why it's not being used. How about:

`true_start = std::clamp(HandleNegativeAxis(true_start, rank), 0, rank)`

Reply:
Thanks @fdwr, updated.